zerver/tests/test_subs.py

```diff
@@ -2198,6 +2204,86 @@ def test_unarchive_stream(self) -> None:
+        """
+        Test permissions for archiving and unarchiving streams, and ensure users without
+        the necessary permissions cannot archive or unarchive a stream.
+        """
+        desdemona = self.example_user("desdemona")
+
+        iago = self.example_user("iago")
+
+        self.login_user(desdemona)
+        stream = self.make_stream("test_stream", invite_only=False)
+        self.subscribe(iago, stream.name)
+
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_success(result)
+        stream.refresh_from_db()
+        self.assertTrue(stream.deactivated)
+
+        result = self.client_patch(
+            f"/json/streams/{stream.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_success(result)
+
+        cordellia = self.example_user("cordelia")
+        self.login_user(cordellia)
+
+        stream1 = self.make_stream("test_stream_1", invite_only=False)
+        self.subscribe(iago, stream1.name)
+
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+        stream1.refresh_from_db()
+        self.assertTrue(stream.deactivated)
+
+        result = self.client_patch(
+            f"/json/streams/{stream1.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+
+        aaron = self.example_user("aaron")
+        self.login_user(aaron)
+
+        self.subscribe(aaron, stream.name)
+
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+
+        result = self.client_patch(
+            f"/json/streams/{stream.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+
+        self.subscribe(aaron, stream1.name)
+        result = self.client_delete(f"/json/streams/{stream1.id}")
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+        self.assertFalse(stream1.deactivated)
+
+        result = self.client_patch(
+            f"/json/streams/{stream1.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+
+    def test_is_archived_true_does_not_archive_stream(self) -> None:
+        """
+        Ensure that passing `is_archived` as True does not archive the stream.
+        """
+        hamlet = self.example_user("hamlet")
+        self.login_user(hamlet)
+        stream = self.make_stream("test_stream", invite_only=False)
+        self.subscribe(hamlet, stream.name)
+
+        do_change_user_role(hamlet, UserProfile.ROLE_REALM_ADMINISTRATOR, acting_user=None)
```

Comment:
Here also you can use admin user directly. I assume you will make this change in one another test as well where role is being changed.